### PR TITLE
Add support for linking PRs

### PR DESCRIPTION
**Pull request by sidpremkumar**
Adds support for linking PRs in Sync2Jira (adds them as a comment downstream)

### CODE REVIEW

**Comment by sidpremkumar** [comment]
Tests have been updated and everything is working. Once Ralph gives the final check we can do a sanity check on all the `TODOs` and make sure everything is working 100% :)

**Comment by sidpremkumar** [comment]
Just finished updating everything including the readthedocs. Would like to do one *final* check once this PR has been reviewed.

File: fedmsg.d/sync2jira.py [hunk 1]
```diff
@@ -29,36 +29,36 @@
...
         # Don't actually make changes to JIRA...
-        'testing': True,
+        'testing': False,
 
         # Set to True when developing to disable sentinel query
         'develop': False,
 
         # Your Github token
-        'github_token': 'YOUR_TOKEN',
+        'github_token': 'c8b9854a8906e5dfa55327fcd209af0b895b1ff0',
```
Comment (ralphbean) [inline fedmsg.d/sync2jira.py]: Uh oh!  :taco:

Remove this from the commit *and* retire that token.  It has been exposed now.

Comment (sidpremkumar) [reply fedmsg.d/sync2jira.py]: super sorry about that, reverting rn

File: sync2jira/downstream_pr.py [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+    # Create a client connection for this issue
+    client = d_issue.get_jira_client(pr, config)
+
+    # Check the status of the JIRA client
+    if not config['sync2jira']['develop'] and not d_issue.check_jira_status(client):
+        log.warning('The JIRA server looks like its down. Shutting down...')
+        raise JIRAError
+
+    # Find our JIRA issue if one exists
+    if isinstance(pr, Issue):
```
Comment (ralphbean) [inline sync2jira/downstream_pr.py]: The docstring says this should be a `PR` object.  It cannot be an `Issue`?

File: sync2jira/downstream_pr.py [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        pr.jira = matcher(pr.content, pr.comments)
```
Comment (ralphbean) [inline sync2jira/downstream_pr.py]: Suggestion, name this `pr.jira_key ` to be clear about what it is.

File: sync2jira/main.py [hunk 4]
```diff
@@ -49,32 +51,43 @@
...
-    'pagure.issue.tag.added': u.handle_pagure_message,
-    'pagure.issue.comment.added': u.handle_pagure_message,
-    'pagure.issue.comment.edited': u.handle_pagure_message,
-    'pagure.issue.assigned.added': u.handle_pagure_message,
-    'pagure.issue.assigned.reset': u.handle_pagure_message,
-    'pagure.issue.edit': u.handle_pagure_message,
-    'pagure.issue.drop': u.handle_pagure_message,
-    'pagure.issue.tag.removed': u.handle_pagure_message,
+# Issue related handlers
+handlers_issue = {
```
Comment (ralphbean) [inline sync2jira/main.py]: `issue_handlers` reads more naturally than `handlers_issue`, no?

File: sync2jira/main.py [hunk 5]
```diff
@@ -49,32 +51,43 @@
...
+    'pagure.issue.comment.edited': u_issue.handle_pagure_message,
+    'pagure.issue.assigned.added': u_issue.handle_pagure_message,
+    'pagure.issue.assigned.reset': u_issue.handle_pagure_message,
+    'pagure.issue.edit': u_issue.handle_pagure_message,
+    'pagure.issue.drop': u_issue.handle_pagure_message,
+    'pagure.issue.tag.removed': u_issue.handle_pagure_message,
+}
+
+# PR related handlers
+handlers_pr = {
```
Comment (ralphbean) [inline sync2jira/main.py]: Same here.

File: sync2jira/downstream_pr.py (hunk 2 above)
Comment (sidpremkumar) [reply sync2jira/downstream_pr.py]: Updated the docstring. The thing is Github pushes all comments (issues and prs) under the suffix 'issue'. So you cannot tell the difference when creating the intermediary object (see: ba2671ffc6056c562689789f47cebf98868c3744). This line is used to add the `.jira` attribute to the issue object. (`.jira_key` now)

File: sync2jira/main.py (hunk 4 above)
Comment (sidpremkumar) [reply sync2jira/main.py]: WFM

File: sync2jira/downstream_pr.py (hunk 3 above)
Comment (sidpremkumar) [reply sync2jira/downstream_pr.py]: WFM

File: sync2jira/downstream_pr.py [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    new_comment = format_comment(pr, pr.suffix)
+    # Check if the comment if already there
+    if not comment_exists(client, existing, new_comment):
+        log.info(f"Added comment for PR {pr.title} on JIRA {pr.jira}")
+        client.add_comment(existing, new_comment)
+
+
+def sync_with_jira(pr, config):
+    """
+    Attempts to sync a upstream issue with JIRA (i.e. by finding
```
[the remark this answers is not in the document]
Comment (sidpremkumar) [reply sync2jira/downstream_pr.py]: Yep

**Pull request merged**